Sources/Moho/ScopeCheck.swift

```diff
@@ -87,51 +89,96 @@ extension NameBinding {
       return self.underScope { _ in
         let bindings = self.scopeCheckBindingList(syntax.bindingList)
         let ee = self.scopeCheckExpr(syntax.bodyExpr)
-        return bindings.reversed().reduce(ee) { (acc, next) in
-          return .lambda(next, acc)
+        return bindings.reversed().reduce(ee) { (acc, next) -> Expr in
+          let boundNames: [Name] = next.0
+          let nameTy: Expr = next.1
+          return boundNames.dropFirst().reduce(nameTy) { (acc, nm) -> Expr in
+            return Expr.lambda((nm, nameTy), acc)
+          }
         }
       }
-    case let syntax as ApplicationExprSyntax:
+    case let syntax as ReparsedApplicationExprSyntax:
       guard syntax.exprs.count > 1 else {
         return self.scopeCheckExpr(syntax.exprs[0])
       }
 
-      guard let headExpr = syntax.exprs[0] as? NamedBasicExprSyntax else {
+      switch syntax.exprs[0] {
+      case let headExpr as NamedBasicExprSyntax:
+        let n = QualifiedName(ast: headExpr.name)
+        guard n.string != "->" else {
+          assert(syntax.exprs.count == 3)
+          return .function(self.scopeCheckExpr(syntax.exprs[1]),
+                           self.scopeCheckExpr(syntax.exprs[2]))
+        }
+
+        var args = [Expr]()
+        for e in syntax.exprs.dropFirst() {
+          let elim = self.scopeCheckExpr(e)
+          args.append(elim)
+        }
+
+        let head: ApplyHead
+        if self.isBoundVariable(n.name) {
+          head = .variable(n.name)
+        } else if let (fqn, _) = self.lookupLocalName(n.name) {
+          head = .definition(fqn)
+        } else if self.lookupFullyQualifiedName(n) != nil {
+          head = .definition(n)
+        } else {
+          // If it's not a definition or a local variable, it's undefined.
+          // Recover by introducing a local variable binding anyways.
+          self.engine.diagnose(.undeclaredIdentifier(n), node: n.node)
+          head = .variable(n.name)
+        }
+
+        return .apply(head, args.map(Elimination.apply))
+
+      default:
         fatalError("Cannot yet handle this case")
       }
-
-      let n = QualifiedName(ast: headExpr.name)
-      guard n.string != "->" else {
-        assert(syntax.exprs.count == 3)
-        return .function(self.scopeCheckExpr(syntax.exprs[1]),
-                         self.scopeCheckExpr(syntax.exprs[2]))
+    case let syntax as ApplicationExprSyntax:
+      guard syntax.exprs.count > 1 else {
+        return self.scopeCheckExpr(syntax.exprs[0])
       }
 
-      var args = [Expr]()
-      for e in syntax.exprs.dropFirst() {
-        let elim = self.scopeCheckExpr(e)
-        args.append(elim)
-      }
+      switch syntax.exprs[0] {
+      case let headExpr as NamedBasicExprSyntax:
+        let n = QualifiedName(ast: headExpr.name)
+        guard n.string != "->" else {
```

Comment:
Would be nice to add an `isArrow` property...

Reply:
It would be nice to delete this monstrosity and do it correctly...

Reply:
👍👏👍👏👍👏👍👏👍👏